### DISCUSSION TITLE
do-while-else statement

### DISCUSSION THREAD

**Discussion by gregsdennis** [dotnet/csharplang#2955] [Language Ideas]
# Problem

I commonly see variants of the following loop:

```c#
while (true) {
    a = random.Next(1, 10);
    b = random.Next(1, 10);

    if (a < b) break;

    Console.WriteLine($"Not yet: {a} {b}");
}
```

This doesn't truly express the intent of the loop.  Here the user wants to

1. perform some action
1. verify the results
1. if verified, break out of the loop
1. otherwise do something else before starting over

Because we want to enter the loop before doing any validation, it is suggested that we use a do-while loop instead:

```c#
do {
    a = random.Next(1, 10);
    b = random.Next(1, 10);

    if (a < b) break;

    Console.WriteLine($"Not yet: {a} {b}");
} while (true);
```

Better, but there is still the matter of the `while (true)` which could be considered to be a code smell.  Ideally, we'd want to have the break-out condition as part of the `while`.

# Proposal

I propose we introduce a *do-while-else* construct that moves the `while` keyword into the middle and introduces a false clause that isn't executed until the condition fails.

```c#
do {
    a = random.Next(1, 10);
    b = random.Next(1, 10);
} while (a < b) else {
    Console.WriteLine($"Not yet: {a} {b}");
}
```

This says

1. calculate two random numbers, `a` and `b`
1. if `a < b`, exit the loop
1. otherwise print that the condition isn't met and start over

# Issues

## Readability

It can be confusing that the `else` portion is part of the loop since the construct is broken into two code blocks.

## Similar constructs in other languages

Python has a [*while-else* construct](https://www.stev.org/post/pythonwhileelsestatement) that operates a bit differently:

```python
while x:
    # do some things
    break
else:
    # do something else
```

This operates in a similar way to an *if-else* except that it stays in the *if* portion while the condition is true.  The C# analog would be this:

```c#
if (x) {
    while (x) {
        // do some things
    }
} else {
    // do something else
}
```

This is **NOT** what I'm proposing.

**Comment by evgenidb** [comment]
> ```cs
> do {
>     a = random.Next(1, 10);
>     b = random.Next(1, 10);
> } while (a < b) else {
>     Console.WriteLine($"Not yet: {a} {b}");
> }
> ```
> 
> This says
> 
>     1. calculate two random numbers, `a` and `b`
> 
>     2. if `a < b`, exit the loop
> 
>     3. otherwise print that the condition isn't met and start over
Not really. It seems to me
1. calculate two random numbers, `a` and `b`
2. if `a < b`, repeat the loop
3. when `a < b` is false, exit the loop and execute the else block.

Besides, this way it's way more unclear than before: the check is still somewhere in the middle, not clear that the loop should break, and the previous version is still more readable and easier to reason.

> Python has a [_while-else_ construct](https://www.stev.org/post/pythonwhileelsestatement) that operates a bit differently:
> 
> ```python
> while x:
>     # do some things
>     break
> else:
>     # do something else
> ```
> 
> This operates in a similar way to an _if-else_ except that it stays in the _if_ portion while the condition is true. The C# analog would be this:
> 
> ```cs
> if (x) {
>     while (x) {
>         // do some things
>     }
> } else {
>     // do something else
> }
> ```
This is false. The else block in while-else always executes when the condition in while becomes false. It could be on the 1st evaluation or on the 100th. The only way to skip it is if you break the loop before the evaluation: break, return from the method, throw an error. That's the reason there is break at the end of the while in your Python example... which makes the while into just an if. It's even explained in the article you linked for while-else (where they explain why break is needed).
So for the example you've given (and in the linked article)
> ```python
> while x:
>     # do some things
>     break
> else:
>     # do something else
> ```
is the same as
```python
if x:
    # do some things
else:
    # do something else
```
Also while-else in Python is confusing. And rarely used. I don't know how many people of those who know of it or still remember of its existence realize that else could be executed not only on the very first evaluation but on any subsequent.
The canonical
```python
while x:
    # do some things
else:
    # do something else
```
will be closer to
```cs
while (true) {
    if (x) {
        // do some things
    }
    else {
        // do something else
        break;
    }
}
```
Confusing, right?

**Comment by Kairus101** [comment]
While I agree that it improves readability in the sense that it better represents logic in reading terms - I find the same scope being shared across the two code blocks extremely unnerving, though I can't think of a way to represent this logic flow without it.

**Comment by ufcpp** [comment]
very similar to https://github.com/dotnet/csharplang/issues/1025#issuecomment-475450644

**Comment by gregsdennis** [comment]
@evgenidb thanks for the python clarification.

---

I agree that this is confusing.  But I see the pattern a lot, and this is the best way I could deal with the redundancy of the `if` and the `while`.

**Comment by gregsdennis** [comment]
@ufcpp good find! That *is* quite similar.

**Comment by vladd** [comment]
What are the proposed scoping rules? Is this allowed to compile:

```csharp
do {
    var a = random.Next(1, 10);
    var b = random.Next(1, 10);
} while (a < b) else { // are a and b available in this line?
    Console.WriteLine($"Not yet: {a} {b}"); // and this?
}
```
?

**Comment by theunrepentantgeek** [comment]
The desire for a "loop-and-a-half" construct is common, it's been a desire of programmers for a very long time. (I recall it coming up during my Computer Science degree in the early 1990's.)

Are there any languages that actually have a _good syntax_ for this that C# can use for inspiration?

Having the **else** part _outside the block_ when it's part of the loop seems really confusing.

**Comment by Joe4evr** [comment]
> Having the **else** part _outside the block_ when it's part of the loop seems really confusing.

Agreed, and I don't believe there's any other construct in C# currently that does this (the closest thing I can think of is the implicit spilling of `out var` variables, but that's a well-defined quirk).

**Comment by gregsdennis** [comment]
@vladd yes, I would expect that the scope is shared between the two blocks, which *does* add to the confusion.

To address this, I had considered a few variations

```c#
do {
    a = random.Next(1, 10);
    b = random.Next(1, 10);
    while (a < b) continue;
    Console.WriteLine($"Not yet: {a} {b}");
}
```

where the loop continues upon meeting the condition, but in this case, `while (a<b) continue;` is valid anyway and means something else.

Another variation is to keep the `else`

```c#
do {
    a = random.Next(1, 10);
    b = random.Next(1, 10);
    while (a < b) else
    Console.WriteLine($"Not yet: {a} {b}");
}
```

The primary issue with these is that there is no clear relationship between the `do` (which is *outside* of the block) and the `while` (which is *inside* the block).

----------------

🤔 Maybe it doesn't need to share scope.  `if-else` doesn't.  In that case, we'd have to declare the variables prior to entering the loop.

```
int a, b;
do {
    a = random.Next(1, 10);
    b = random.Next(1, 10);
} while (a < b) else {
    Console.WriteLine($"Not yet: {a} {b}");
}
```

(We don't need to initialize `a` or `b` because they get values inside the loop before they're used in the `else` clause.)

**Comment by gregsdennis** [comment]
@Joe4evr That's a good point about the special case of the `out var` syntax.  Although it *is* well-defined now, it was a point of a lot of questions and debate on how it should work.  Now it's part of the language, and people are familiar with it.  I expect this would be the same.

**Comment by gregsdennis** [comment]
~@theunrepentantgeek but it's not *really* "part of the loop."  It's a tail.  It only executes when the condition fails, exiting the loop.~  Sorry, early morning brain.

**Comment by theunrepentantgeek** [comment]
It's **totally a part of the loop** - it's executed _every time_ the loop executes, except perhaps the last time.

Your original post has both of these as equivalent:

``` csharp
do {
    a = random.Next(1, 10);
    b = random.Next(1, 10);

    if (a < b) break;

    Console.WriteLine($"Not yet: {a} {b}");
} while (true);
```

``` csharp
do {
    a = random.Next(1, 10);
    b = random.Next(1, 10);
} while (a < b) else {
    Console.WriteLine($"Not yet: {a} {b}");
}
```

If the last bit was a tail, the logical way to write it would be to move the `Console.WriteLine()` until after the loop body.

**Comment by gregsdennis** [comment]
Oh... Sorry.  It was early morning for me.  You're right.

**Comment by gregsdennis** [comment]
Ooooorrrrr...  "until" (with the inverse condition)

```c#
do {
    a = random.Next(1, 10);
    b = random.Next(1, 10);
    until (a >= b);
    Console.WriteLine($"Not yet: {a} {b}");
}
```

**Comment by XNTEABDSC** [comment]
how about remove else
```
do {
    var a = random.Next(1, 10);
    var b = random.Next(1, 10);
    } while(a<b) {
    Console.WriteLine($"Not yet: {a} {b}");
}
```
it can be combined with do{}while() ;  \<- considered as empty statement (and change do-while scope rule)
